consult.el

```diff
@@ -1102,46 +1102,81 @@ for which the command history is used."
+                                    (boundp (cadr ring))))
+                             consult-mode-histories))
+              (user-error
+               "No history configured for `%s', see `consult-mode-histories'"
+               major-mode))
+        (cons (symbol-value history)
+              index-var)))))
+
+(defun consult--label-and-delete-dups (entries)
+  "Label ENTRIES with history indices.
+
+If there are duplicates, the most recent occurrence of the
+duplicate gets labeled with its index, and the others are not
+shown."
+  (let ((ht (make-hash-table :test 'equal))
+        (idx 0)
+        entry
+        accum)
+    (while (setq entry (pop entries))
+      (unless (gethash entry ht)
+        (puthash entry t ht)
+        (push (cons entry (cons idx entry))
+              accum))
+      (cl-incf idx))
+    (nreverse accum)))
 
 (defun consult--history-elements (history)
   "Return elements from HISTORY.
 Can handle lists and rings."
-  (delete-dups (seq-copy (if (ring-p history)
-                             (ring-elements history)
-                           history))))
+  (consult--label-and-delete-dups (seq-copy (if (ring-p history)
+                                                (ring-elements history)
+                                              history))))
 
 ;; This command has been adopted from https://github.com/oantolin/completing-history/.
 ;;;###autoload
 (defun consult-history (&optional history)
   "Insert string from buffer HISTORY."
   (interactive)
   (let* ((enable-recursive-minibuffers t)
-         (str (consult--read "History: "
-                             (or (consult--history-elements
-                                  (or history (consult--current-history)))
-                                 (user-error "History is empty"))
-                             :history t ;; disable history
-                             :sort nil)))
+         (dup-cands-and-index-var (if history
+                                      (cons history 'nil)
+                                    (consult--current-history)))
+         (dup-cands (car dup-cands-and-index-var))
+         (index-var (cdr dup-cands-and-index-var))
+         (unique-cands-with-indices (consult--history-elements dup-cands))
+         (selected-cand (consult--read "History: "
+                                       (or unique-cands-with-indices
+                                           (user-error "History is empty"))
+                                       :history t ;; disable history
+                                       :lookup #'consult--lookup-candidate
+                                       :sort nil))
+         (index-of-selected-candidate (car selected-cand))
+         (selected-str (cdr selected-cand)))
     (when (minibufferp)
       (delete-minibuffer-contents))
-    (insert (substring-no-properties str))))
+    (pcase index-var
+      ((or 'comint-input-ring-index
+           'eshell-history-index)
```

Comment:
Why is this hard-coded?

Reply:
I didn't see a better way of handling this. `index-var` can in general be `'nil`, corresponding to modes that don't keep track of any kind of index. Do you have any suggestions?

Reply:
I don't understand. If there is an index (non-nil) update it, otherwise do nothing? There must not be any hard coding since otherwise this won't work if the user customizes the history list.

Reply:
`minibuffer-history-position` seems to operate differently, so at minimum we'd have to special-case that.

Would this be better?

``` elisp
(pcase index-var
      ('nil
       nil)
      ('minibuffer-history-position
       (setf (symbol-value index-var) (1+ index-of-selected-candidate)))
      (_
       (setf (symbol-value index-var) index-of-selected-candidate)))
```

Reply:
We already special case the minibuffer for some other reasons, right? So this is justified.